### PR TITLE
Media & Text block: Remove the link option when the featured image is used

### LINKED ISSUE #60497: Media & Text block: The link option does not work when the featured image is used
### Description

When the Media & Text block uses the featured image, and a link is added using the block toolbar option, the link does not work because it does not wrap around the image.

[It is not added to the block in index.php.](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/media-text/index.php#L32)

### Step-by-step reproduction instructions

Activate Gutenberg
Create a new post and add a featured image.
Add a Media & Text block and select the option "Use featured image".
In the block toolbar, select the link option and add a link.
Save and view the front.
Confirm that the image is not linked.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

Gutenberg 18.0 or newer, or Gutenberg trunk.

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by carolinan**
## What & Why & How?

When the Media & Text block is linked and uses the featured image, the image is not linked on the front of the site.

Originally, this PR was meant to make the link work.
The preferred way to solve the bug, as expressed in the reviews below, would be to update the LinkControl component and replace ImageURLInputUI.

It was then recommended to solve the bug short-term by not showing the link option if the featured image is used, so that
the user can not use the broken control.
The PR was updated to reflect this.

## Updated testing instructions:

1. Activate Gutenberg
2. Create a new post and add a featured image.
3. Add a Media & Text block and select the option "Use featured image".
4. In the block toolbar, confirm that there is no link option.
5. Select another image from the media library.
6. In the block toolbar, confirm that there is a link option, set a link. Save and view the front, confirm that the link is working.
7. Select the featured image again, save and view the front. confirm that there is no link being output. This step is meant to to test that the link that was set in step 6 was properly unset (reset) when the featured image was selected.

<details><summary> Old description</summary>
The pull request updates where in the block markup the `img` tag is inserted:

1. It checks wether the block `$content` contains an `<a>` tag.
2. If there is an `<a>` tag, it inserts the  `` tag by replacing the closing `</a>` tag with  `</a>`.
By replacing the closing `</a>` instead of the opening tag, the class names and attributes on the link are preserved.
3. If the block `$content` does not contain a link, the `` tag is inserted directly inside the `<figure>`.

Closes https://github.com/WordPress/gutenberg/issues/60497

## Limitations
The PR does not address the problem that users can not set _individual_ links on the block when it is placed inside a query loop.

## Testing Instructions

1. Activate Gutenberg
2. Create a new post and add a featured image.
3. Add a Media & Text block and select the option "Use featured image".
4. In the block toolbar, select the link option and add a link.
5. Save and view the front.
6. Confirm that the image is linked.
</details>

### CODE REVIEW

**Comment by carolinan** [comment]
With this PR, if the featured image is linked but has no alt text, VoiceOver on Mac will read the URL is the link text.

I think that is the best that can be done unless we want to add the "Link to post" setting (the same setting that the featured image block has).

**Review by t-hamano — COMMENTED** [review]
Thanks for the PR!

I think the difficult part of this problem is finding the link of the featured image when the block contains multiple `a` elements.

For example, when "Show media on the right" is enabled and a link exists in the content, the screenshot and the markup will look like this:

image

```html
<!-- wp:media-text {"mediaPosition":"right","linkDestination":"custom","mediaType":"image","imageFill":false,"useFeaturedImage":true} -->
<div class="wp-block-media-text has-media-on-the-right is-stacked-on-mobile">
	<div class="wp-block-media-text__content">
		<!-- wp:paragraph {"placeholder":"Content…"} -->
		<p>Hello <a href="#">World</a></p>
		<!-- /wp:paragraph -->
	</div>
	<figure class="wp-block-media-text__media">
		<a href="#test"></a>
	</figure>
</div>
<!-- /wp:media-text -->
```

On the front end side, the following invalid markup is output:

```html
<div class="wp-block-media-text has-media-on-the-right is-stacked-on-mobile">
	<div class="wp-block-media-text__content">
	  <p>Hello 
	    <a href="#">World
	      <img decoding="async" width="1024" height="683" alt="" class="wp-image-8 size-full" src="http://localhost:8888/wp-content/uploads/2024/04/test_1.jpg" srcset="http://localhost:8888/wp-content/uploads/2024/04/test_1.jpg 1024w, http://localhost:8888/wp-content/uploads/2024/04/test_1-300x200.jpg 300w, http://localhost:8888/wp-content/uploads/2024/04/test_1-768x512.jpg 768w" sizes="(max-width: 1024px) 100vw, 1024px" />
	    </a>
	  </p>
	</div>
	<figure class="wp-block-media-text__media">
	  <a href="#test">
	    <img>
	  </a>
	</figure>
</div>

```

I'm wondering how many users will want to take advantage of the custom link when they enable the featured image on this block. For example, when this block is inserted inside a Query Loop block and the link is enabled, I expect the featured image to automatically link to that post.

In any case, we need to find a way to detect the correct link. Ideally, it would be nice if we could specify the `a` element inside the `.wp-block-media-text__media` class, but as far as I know, `WP_HTML_Tag_Processor` does not allow us to specify the parent element 😅

**Comment by carolinan** [comment]
Thank you for lending me your brain!
I'll try this option:
> I think that is the best that can be done unless we want to add the "Link to post" setting (the same setting that the featured image block has).

**Comment by carolinan** [comment]
- The link control in the toolbar is hidden when the featured image is selected.
- A new toggle option in the sidebar, "link to (Post type name)" is available when the featured image is selected.
- There is a new attribute called "featuredImageLink" that is `true` when the option is enabled. This name could be improved.

index.php:
- Use the post title as the alternative text when the featured image is linked, if a post title is set.
- Use the alternative text from the media library if the image is not linked and the text is set.
- Use the `WP_HTML_Tag_Processor` to find the HTML tag with the class name `wp-block-media-text__media` instead of finding just any `<figure>` tag in the block content.
This is because the content part of the block can include an image block that also uses a `figure` element, and the src and class names etc would override the first image block. 🙀 
There may be more scenarios like that, that I have not found yet.

### What is still missing:

While I could add conditional options for changing the `linkTarget` and `rel`,  when the link to post option is enabled, I was not able to figure out how to pass those two attributes to the `index.php` file.
They don't show up in `$attributes` so they were probably not updated correctly.

**Comment by t-hamano** [comment]
Thanks for the update!

> While I could add conditional options for changing the `linkTarget` and `rel`,  when the link to post option is enabled, I was not able to figure out how to pass those two attributes to the `index.php` file.
> They don't show up in `$attributes` so they were probably not updated correctly.

This is because those attributes are not saved as HTML comments, but [as part of the block HTML](https://github.com/WordPress/gutenberg/blob/7061bf815ac8d885d5923fdb6808f098b15ea701/packages/block-library/src/media-text/block.json#L48-L53). If you remove the `source`, `selector`, and `attribute` from those attributes, the data will be saved as a HTML comment and can be referenced in the `render_callback()` function. If ywe do this we will need to add a deprecation. Alternatively, we could parse the media link element using the HTML Tag Processor and retrieve the attributes via `get_attribute()` mothod.

---

BTW, if I think about this problem again, it may be more difficult than I imagine.

@draganescu @aaronrobertshaw I have summarized the problems and issues that I understand below, so if you have a good approach, please let me know 🙏

- Originally reported in [#60497](https://github.com/WordPress/gutenberg/issues/60497), an issue where the custom link was not applied correctly on the front end when "Use Featured Image" was enabled for the Media & text block.
- In this PR, if "Use Featured Image" is enabled, the custom link will be disabled and a "Link to Post" control will be added to the block sidebar. This is because the `ImageURLInputUI` component used in this block does not accept custom settings.
- The custom link supports "Open in new tab", "Link Rel", and "Link CSS Class" in addition to the URL. These UIs need to be added to the block sidebar so that these settings can still be controlled when the featured image is linked to a post.

Considering these, I think the UI will probably look something like the following, what do you think? I'm concerned that the UI location for setting the link changes depending on whether the featured image is enabled or disabled.


https://github.com/WordPress/gutenberg/assets/54422211/3fe10f62-ef0c-464c-a10a-e5ac6f6c2096

**Comment by carolinan** [comment]
I have the same concerns.
There is more though, - what if we eventually want to add the lightbox? It will be even more confusing with separate settings.

**Comment by carolinan** [comment]
I opened this issue, but I'm not sure I described the problem clearly enough:
Link control: Make it possible to link to the current post when in a query loop
https://github.com/WordPress/gutenberg/issues/60499

**Comment by carolinan** [comment]
...Should just have made the block with an inner media block... 🙄

**Comment by carolinan** [comment]
If all blocks that use the featured image link option could move it to the link setting in the toolbar, it would be more consistent with how the "lightbox linking exception" works. I mean the exception where you can only choose one link or the other.

But short term, what can we do to fix this rather large bug as soon as possible?

**Comment by carolinan** [comment]
Pinging @WordPress/gutenberg-design for feedback.
Summary:
The Media & Text block is the only media block that can be linked to: anything, the media attachment page, the media file, and conditionally, to the post or page that the featured image is attached to.

Short term, there is a bug that needs to be fixed, that is present in the released version of Gutenberg, where linking the Media & Text block when the featured image option is selected, does not work on the front.

Long term, we may need to rethink and improve the UX of how we link a featured image to its post. Right now blocks that uses featured images has a toggle option in the sidebar. This is the same kind of toggle option that is used to enable linking the post title or site title.

But, this option becomes inconsistent when the block **also** has a link control in the block toolbar:
-As in the proposed change in this PR
-As when the lightbox option was introduced and eventually moved to the link control in the toolbar

So we need to decide:
 - If a short term solution to fix the bug is acceptable even if we consider the UX to be less than ideal?
 - Wether the link option should be consistent with how the other featured image blocks, **or** with how the lightbox option works.
 - Wether all link enabling options should actually be moved away from the sidebar, for all these blocks.
 
This is not the correct place to discuss the long term solutions, and sharing of screenshots etc should be moved to a separate issue. I just feel strongly that the current bug needs to be fixed, since I caused it, and also I want to use the block 😅

**Comment by jasmussen** [comment]
Thanks for the ping. I'm happy to move this to a new issue if you prefer, but ultimately the right solution here is the one you allude to, unifying all linking to the link dropdown. That is, in this dropdown, there could be an option saying "Link to post":

media and text link dropdown

That option could be present only when the featured image is set. When linked, it would show a similar state to the current options:

media and text linked dropdown

So in the above context, it could say "Link to post ↗".

**Comment by draganescu** [comment]
I agree w/ @jasmussen about definitely not moving linking to the inspector - I think it would be a very isolated UX (an exception from how we generally link). I would also add that if we link to post what is the point of rel and target which make sense for external links?

**Comment by richtabor** [comment]
In https://github.com/WordPress/gutenberg/issues/50893 I was thinking how link editing would consolidate across WordPress. 

It could be appended to the list in the link UI as pictured above, but I'd like to move forward on using LinkControl throughout the experience, to reduce ad hoc fixes/implementations.

**Comment by carolinan** [comment]
Yes I have intentionally avoided creating yet another custom and temporary link control.

I still do think that this bug needs to be fixed and not get blocked.

**Comment by carolinan** [comment]
Is this solution acceptable until someone is able to re-work the LinkControl? What other options are there besides disabling the link settings or reverting the featured image option? I don't want this to go broken into 6.6.

**Comment by carolinan** [comment]
If the link option is disabled when the featured image is used, using the featured image in the media and text block inside a query is kind of pointless. 
It would still need to manage the broken links already added by users who have had Gutenberg active.

**Review by draganescu — COMMENTED** [review]
Given the feedback that ideally we move everything to the linking UI and since this is a bug fix what do y'all think if we just hide _for now_ the linking option if the featured media image is used in the media text block?

I understand this reduces the usefulness of the featured image prop in the media text block, but it'd be a temporary solution that avoids yet another deprecation, and hundreds of documents updated when linking will happen in the toolbar.

I say this PR should only fix the bug and we then open a separate issue and PR to introduce link to post to link UI?

File: packages/block-library/src/media-text/edit.js [hunk 1]
```diff
@@ -282,6 +294,7 @@ function MediaTextEdit( {
 					focalPoint: undefined,
 					mediaWidth: 50,
 					mediaSizeSlug: undefined,
+					featuredImageLink: false,
```
Comment (draganescu) [inline packages/block-library/src/media-text/edit.js]: I wonder if this should be toggled to false if this flow happens:

- set some image
- set featured image
- set link to post type on
- set some other image
- set featured image
- should link to post type be off now?

Comment (carolinan) [reply packages/block-library/src/media-text/edit.js]: Yeah it should be off I think...

Comment (carolinan) [reply packages/block-library/src/media-text/edit.js]: But it wont matter I guess for this PR if I disable the link option.

**Comment by carolinan** [comment]
I'm not fully convinced that having the link option in the sidebar is more disruptive than not having a link option at all, but since this is the only feedback received I will try to update the PR so that this can be solved, short term.

**Comment by draganescu** [comment]
@carolinan I am not convinced either 100% 😁  but I lean towards removing linking altogether because the featured image prop of the media text block is new (will land in WP next), and we found a limitation of our linking system, so I generally default to less instead of more to undo later.

**Comment by carolinan** [comment]
I still want to test that there is no "weirdness" for people who has used Gutenberg and added links.
For those users, the block attributes related to the link won't update unless they toggle the featured image off...

**Comment by carolinan** [comment]
For blocks that had the link set before this PR, the link is still output in the wrong position on the front:
```
<div class="wp-block-media-text is-stacked-on-mobile"><figure class="wp-block-media-text__media">
<img decoding="async" width="295" height="295" alt="" class="wp-image-201 size-full" src="..." 
sizes="(max-width: 295px) 100vw, 295px"><a href="https://wordpress.org"></a></figure>
<div class="wp-block-media-text__content">
<p>This Media &amp; Text block was linked on Gutenberg trunk.</p>
</div></div>
```

Do I need to try to remove it?

**Comment by draganescu** [comment]
> For blocks that had the link set before this PR, the link is still output in the wrong position on the front [...] Do I need to try to remove it?

I would lean into no - using the plugin in production is precisely subject to these kinds of issues. #51491 introduced using the featured image for the media and text block which is a feature only available starting WP 6.6.

If the bug landed in a WP release (minor or major) a minor release to remove this from where it was output (via some filter) would be justified. But since the bug existed only in the plugin, adding some logic to strip that out would be then shipped forever for a tiny fraction of cases.

cc @Mamaduka is my reasoning above correct?

**Comment by Mamaduka** [comment]
@draganescu, yes, the reasoning makes sense. There is no need to introduce backward compatibility handers for code that has only ever existed in plugins.

Also, if the content had been saved in DB, it would've been a different story, but IIRC, markup is dynamic for featured images.

**Review by draganescu — APPROVED** [review]
This LGTM. Let's make an issue if it's not made to add "link to current post" the link UI somehow.

**Pull request merged**